src/Storages/MergeTree/MergeTask.cpp

```diff
@@ -518,7 +517,23 @@ bool MergeTask::VerticalMergeStage::prepareVerticalMergeForAllColumns() const
             "of bytes written to rows_sources file ({}). It is a bug.",
             sum_input_rows_exact, input_rows_filtered, cleanedup_rows_count, rows_sources_count);
 
-    ctx->rows_sources_read_buf = std::make_unique<CompressedReadBufferFromFile>(ctx->tmp_disk->readFile(fileName(ctx->rows_sources_file->path())));
+    /// TemporaryDataOnDisk::createRawStream returns WriteBufferFromFile implementing IReadableWriteBuffer
+    /// and we expect to get ReadBufferFromFile here.
+    /// So, it's relatively safe to use dynamic_cast here and downcast to ReadBufferFromFile.
+    auto * wbuf_readable = dynamic_cast<IReadableWriteBuffer *>(ctx->rows_sources_uncompressed_write_buf.get());
+    auto reread_buf = wbuf_readable ? wbuf_readable->tryGetReadBuffer() : nullptr;
+    if (!reread_buf)
+        throw Exception(ErrorCodes::LOGICAL_ERROR, "Cannot reread temporary file {}", ctx->rows_sources_uncompressed_write_buf->getFileName());
+    auto * reread_buffer_raw = dynamic_cast<ReadBufferFromFile *>(reread_buf.get());
+    if (!reread_buffer_raw)
+        throw Exception(ErrorCodes::LOGICAL_ERROR, "Cannot reread temporary file {}", ctx->rows_sources_uncompressed_write_buf->getFileName());
```

Comment:
May be makes sense to change exception message a bit so that this one and the exception above one are not identical -- to be able to tell where exactly the exception was thrown.